docs/proposals/env-vars.md

```diff
@@ -0,0 +1,315 @@
+	// Cannot be updated.
+	// +optional
+	EnvFrom []EnvFromSource `json:"envFrom,omitempty" protobuf:"bytes,19,rep,name=envFrom"`
+	// List of environment variables to set in the container.
+	// Cannot be updated.
+	// +optional
+	// +patchMergeKey=name
+	// +patchStrategy=merge
+	Env []EnvVar `json:"env,omitempty" patchStrategy:"merge" patchMergeKey:"name" protobuf:"bytes,7,rep,name=env"`
+```
+
+First decision:  do we need to expose both `EnvFrom` and `EnvVar` off of `Containers`, especially given that
+`EnvVar` has an optional `EnvFromSource` [ref](https://github.com/kubernetes/api/blob/v0.21.0/core/v1/types.go#L1929)
+
+```go
+// EnvVar represents an environment variable present in a Container.
+type EnvVar struct {
+	// Name of the environment variable. Must be a C_IDENTIFIER.
+	Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
+
+	// Optional: no more than one of the following may be specified.
+
+	// Variable references $(VAR_NAME) are expanded
+	// using the previous defined environment variables in the container and
+	// any service environment variables. If a variable cannot be resolved,
+	// the reference in the input string will be unchanged. The $(VAR_NAME)
+	// syntax can be escaped with a double $$, ie: $$(VAR_NAME). Escaped
+	// references will never be expanded, regardless of whether the variable
+	// exists or not.
+	// Defaults to "".
+	// +optional
+	Value string `json:"value,omitempty" protobuf:"bytes,2,opt,name=value"`
+	// Source for the environment variable's value. Cannot be used if value is not empty.
+	// +optional
+	ValueFrom *EnvVarSource `json:"valueFrom,omitempty" protobuf:"bytes,3,opt,name=valueFrom"`
+}
+```
+
+This proposal is making the simplifying assumption that ultimately setting the `Env []EnvVar` field in the 
+underlying Tekton `Step`/`Container` is sufficient for an initial implementation and scenarios we want to support.
+That will help the initial implementation in that the explicit "name" index field helps with default kubernetes 
+"path" API operations that tend to occur with devops/gitops flows.
+
+Next, we'll couple the default value capability available in Tekton's `StepTemplate` with reuse of the k8s
+`EnvVar` type.  In other words, a default value field can be mapped to a `StepTemplate`'s env var setting.  
+
+With that, a new Shipwright type would be:
+
+```go
+type EnvironmentVariable struct {
+   corev1.EnvVar
+   
+   // DefaultValue if set is the value applied if neither EnvVar.Value or EnvVar.ValueFrom have a value even 
+   // after any Parameter variable substitution.
+   // +optional
+   DefaultValue  string
+}
+```
+
+Then, the specs of all `BuildStrategy` (both cluster and namespaced scope), `Build`, and `BuildRun` will be 
+updated with an array of `EnvironmentVariable`
+
+`BuildStrategy` would also have a `EnvironmentVariable` array named `DisallowedEnvironmentVariables` that lists `EnvVar` settings which are not allowed
```

Comment:
Thinking out loud, I wonder if we should invert this logic and let administrators define "allowed environment variables" instead.  Like for instance `HTTP_PROXY`, an administrator would like to leave a list of proxy servers that are allowed; in comparison, would be difficult to add a list of servers that aren't allowed.

Reply:
Yeah I could see both allow and deny lists being desirable @otaviof .... I tried to indicate that providing some form of both may be a item to strive fo, though maybe we stage one before the other wrt to the notion of a staging plan I proposed in response to some of @SaschaSchwarze0 's comments.  Though based on some of the exchanges up above, I have some optimism we devise and order of precedence that gets us suitable allow/deny capabilities.

Bottom line though, I need to add more details to the portion you've highlighted here.